core/sr-io/src/offchain/http.rs

```diff
@@ -0,0 +1,398 @@
+	/// Request list of headers.
+	headers: Vec<(Vec<u8>, Vec<u8>)>,
+}
+
+impl<T: Default> Default for Request<'static, T> {
+	fn default() -> Self {
+		Request {
+			method: Method::Get,
+			url: "http://localhost",
+			headers: Vec::new(),
+			body: Default::default(),
+			deadline: None,
+		}
+	}
+}
+
+impl<'a, T: Default> Request<'a, T> {
+	/// Create new Request builder with given URL.
+	pub fn new(url: &'a str) -> Self {
+		let mut req = Request::default();
+		req.url(url);
+		req
+	}
+
+	/// Change the URL of the request.
+	pub fn url(&mut self, url: &'a str) -> &mut Self {
+		self.url = url;
+		self
+	}
+
+	/// Set the body of the request.
+	pub fn body(&mut self, body: T) -> &mut Self {
+		self.body = body;
+		self
+	}
+
+	/// Add a header.
+	pub fn add_header(&mut self, name: &str, value: &str) -> &mut Self {
+		self.headers.push((name.as_bytes().to_vec(), value.as_bytes().to_vec()));
+		self
+	}
+
+	/// Set the deadline of the request.
+	pub fn deadline(&mut self, deadline: Timestamp) -> &mut Self {
+		self.deadline = Some(deadline);
+		self
+	}
+}
+
+impl<'a, 'b, T: IntoIterator<Item=&'b [u8]>> Request<'a, T> {
+	/// Send the request and return a handle.
+	///
+	/// Err is returned in case the deadline is reached
+	/// or the request timeouts.
+	pub fn send(self) -> Result<PendingRequest, ()> {
+		let meta = &[];
+
+		// start an http request.
+		let id = crate::http_request_start(self.method.as_ref(), self.url, meta)?;
+
+		// add custom headers
+		for (header_name, header_value) in &self.headers {
+			crate::http_request_add_header(
+				id,
+				from_utf8(header_name).expect("Header names are always Vecs created from valid str; qed"),
+				from_utf8(header_value).expect("Header values are always Vecs created from valid str; qed"),
+			)?
+		}
+
+		// write body
+		for chunk in self.body {
+			crate::http_request_write_body(id, chunk, self.deadline)?;
+		}
+
+		// finalise the request
+		crate::http_request_write_body(id, &[], self.deadline)?;
+
+		Ok(PendingRequest {
+			id,
+		})
+	}
+}
+
+/// A request error
+pub enum Error {
+	/// Deadline has been reached.
+	DeadlineReached,
+	/// Request had timed out.
+	Timeout,
+	/// Unknown error has been ecountered.
+	Unknown,
+}
+
+/// A struct representing an uncompleted http request.
+#[derive(PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct PendingRequest {
+	/// Request ID
+	pub id: RequestId,
+}
+
+impl PendingRequest {
+	/// Wait for the request to complete.
+	///
+	/// NOTE this waits for the request indefinitely.
+	pub fn wait(self) -> Result<Response, Error> {
+		match self.try_wait(None) {
+			Ok(res) => res,
+			Err(_) => panic!("Since `None` is passed we will never get a deadline error; qed"),
+		}
+	}
+
+	/// Attempts to wait for the request to finish,
+	/// but will return `Err` in case the deadline is reached.
+	pub fn try_wait(self, deadline: impl Into<Option<Timestamp>>) -> Result<Result<Response, Error>, PendingRequest> {
```

Comment:
Nit: you probably know, but nested results doesn't look that awesome. It is hard to wrap my head around which `Err` serves which purpose...